dev/src/bundle.ts

```diff
@@ -41,31 +42,36 @@ export class BundleBuilder {
   constructor(private bundleId: string) {}
 
   add(documentSnapshot: DocumentSnapshot): BundleBuilder;
-  add(queryName: string, querySnap: QuerySnapshot): BundleBuilder;
+  add(queryName: string, querySnapshot: QuerySnapshot): BundleBuilder;
   /**
    * Adds a Firestore document snapshot or query snapshot to the bundle.
    * Both the documents data and the query read time will be included in the bundle.
    *
    * @param {DocumentSnapshot | string=} documentOrName A document snapshot to add or a name of a query.
-   * @param {Query?=} querySnapshot A query snapshot to add to the bundle, if provided.
+   * @param {Query=} querySnapshot A query snapshot to add to the bundle, if provided.
    * @returns {BundleBuilder} This instance.
    *
    * @example
    * let bundle = firestore.bundle('data-bundle');
-   * const bundleStream =
-   *    await bundle.add(await firestore.doc('abc/123').get()) // Add a document
-   *                .add('coll-query', await firestore.collection('coll').get()) // Add a named query.
-   *                .stream();
+   * const docSnapshot = await firestore.doc('abc/123').get();
+   * const querySnapshot = await firestore.collection('coll').get();
+   * const bundleStream = bundle.add(docSnapshot) // Add a document
+   *                            .add('coll-query', querySnapshot) // Add a named query.
+   *                            .stream();
    * // Stream `bundleStream` to clients.
    */
   add(
     documentOrName: DocumentSnapshot | string,
     querySnapshot?: QuerySnapshot
   ): BundleBuilder {
-    if (documentOrName instanceof DocumentSnapshot) {
+    if (
+      documentOrName instanceof DocumentSnapshot &&
+      validateOptional(querySnapshot, {optional: true})
```

Comment:
`validateOptional` only validates that `querySnapshot` can be `undefined` not that it needs to be `undefined` when `documentOrName` is a `DocumentSnapshot`.

This validation should also be moved into the if branch.